# II deo
### Pre same obrade podataka i donošenja zaključaka, neophodno je tzv. "sirovi" (raw) dataset da sredimo tako da bude pogodan za obradu. Jedna od glavnih aktivnosti u okviru tog sređivanja jeste rešavanje problema nedostajućih vrednosti.

In [15]:
import pandas as pd
from pandas import Series

dfr = pd.read_csv('../data/brit.csv')
dfr.sample(8)

Title                        AlbumName  \
421                   Coquillages                    Love in a Mist   
959   2000 Light Years from Home    Their Satanic Majesties Request   
1224           How We Were Before                        I Love You   
873                        Money                 The Rolling Stones   
1083       Don't Forget About Me                  Where Am I Going?   
412                       Lullaby                North Country Maid   
1145            Heinz Baked Beans                  The Who Sell Out   
1192                  Lost Woman                 Roger the Engineer   

                                           Record Label           Performer  \
421                                               Decca  Marianne Faithfull   
959                                       ABKCO Records  The Rolling Stones   
1224                          Varèse Sarabande / Varese         The Zombies   
873                                                 NaN  The Rolling Stones   
1083  Def Jam / Island / Philips / Universal / Virgi...   Dusty Springfield   
412                                           Universal  Marianne Faithfull   
1145                                                MCA             The Who   
1192                                              Edsel       The Yardbirds   

      Track.number Song.duration  \
421             10          3:38   
959              9          4:45   
1224             2          2:05   
873              2          2:31   
1083            14          2:47   
412             11          2:40   
1145             2          1:00   
1192             1          3:14   

                                               Composer  Year  \
421                                 Michael John Taylor  1967   
959                        Mick Jagger / Keith Richards  1967   
1224                                    Colin Blunstone  1966   
873                   Janie Bradford / Berry Gordy, Jr.  1964   
1083                         Gerry Goffin / Carole King  1967   
412                                            Jon Mark  1966   
1145                                     John Entwistle  1967   
1192  Jeff Beck / Chris Dreja / Jim McCarty / Keith ...  1966   

                Genre                                             Styles  ...  \
421          Pop/Rock  British Invasion, Girl Groups, Folk-Rock, Baro...  ...   
959          Pop/Rock  British Psychedelia, Contemporary Pop/Rock, Ps...  ...   
1224         Pop/Rock  British Invasion, Baroque Pop, AM Pop, Contemp...  ...   
873          Pop/Rock  Blues-Rock, British Invasion, Contemporary Pop...  ...   
1083    Pop/Rock, R&B  Blue-Eyed Soul, British Invasion, Pop-Soul, So...  ...   
412          Pop/Rock           British Invasion, Folk-Rock, Baroque Pop  ...   
1145         Pop/Rock  Album Rock, British Invasion, British Psychede...  ...   
1192  Pop/Rock, Blues  Blues-Rock, British Blues, British Invasion, B...  ...   

      speechiness    tempo  valence  duration_ms  popularity  \
421        0.0281   84.755    0.337     218160.0         5.0   
959        0.0490  122.328    0.278     285053.0         0.0   
1224       0.0310  109.043    0.631     125707.0        36.0   
873        0.0424  137.319    0.880     151320.0        18.0   
1083       0.0309  128.049    0.533     167733.0        14.0   
412        0.0413  121.747    0.206     159813.0         0.0   
1145       0.2980  121.300    0.670      60827.0         0.0   
1192          NaN      NaN      NaN          NaN         NaN   

      shake.the.audience  romantic   obscene  family.spiritual   sadness  
421             0.002632  0.111304  0.002632          0.002632  0.002632  
959             0.002506  0.316350  0.002506          0.002506  0.080953  
1224            0.110119  0.001144  0.337479          0.272042  0.001074  
873             0.002552  0.001698  0.002506          0.002506  0.001284  
1083            0.031503  0.159885  0.354789          0.001754  0.001012  
412      

### Najpre ćemo utvrditi koliko nedostajućih (NaN/None) vrednosti ima u svakoj koloni. Metoda isna() preslikava svako polje datafrejma u True/False (ako je polje NaN dobiće vrednost False).

In [3]:
dfr.isna().sum().sort_values(ascending=False)

duration_ms           127
valence               127
tempo                 127
speechiness           127
loudness              127
liveness              127
instrumentalness      127
energy                127
danceability          127
acousticness          127
popularity            127
Song.duration          43
Record Label            3
Composer                0
Track.number            0
Performer               0
Title                   0
AlbumName               0
Styles                  0
Genre                   0
Year                    0
shake.the.audience      0
romantic                0
obscene                 0
family.spiritual        0
sadness                 0
dtype: int64

## 1. varijanta: popunjavanje nedostajućih na osnovu poznatih vrednosti
 ### Nedostajućim vrednostima kolone `loudness` biće dodeljena srednja vrednost te kolone. Nakon popunjavanja nedostajucih vrednosti korišćenjem metode `fillna`, metoda `value_counts` nam vraća broj pojavljivanja svake jedinstvene vrednosti. Srednja vrednost se pojavljuje najčešće, i to kod svih 127 prethodno nedostajućih polja.

In [4]:
dfr['loudness'] = dfr['loudness'].fillna(value=dfr['loudness'].mean())
dfr['loudness'].value_counts()

loudness
-9.566823     127
-9.197000       3
-8.887000       3
-9.436000       3
-10.086000      3
             ... 
-9.310000       1
-9.562000       1
-9.299000       1
-11.992000      1
-9.279000       1
Name: count, Length: 1028, dtype: int64

## 2. varijanta: uklanjanje redova koji sadrže nedostajuće vrednosti
 ### Uklonićemo sve redove kod kojih kolonama energy i tempo nedostaje vrednost, koristeći metodu `dropna`. Kod `subset` parametra navode se kolone za koje se nedostajuće vrednosti razmatraju, dok parametar `how` uzima vrednost `all` (red se briše samo ako su sve vrednosti kolona iz subseta nedostajuće) ili `any` (red se briše ako nedostaje bilo koja vrednost iz subseta). Kako bismo efekat ove naredbe preneli na naš datafrejm, parametar `inplace` postavljamo na True.

In [5]:
dfr.dropna(subset=['energy','tempo'],how='all', inplace=True)
dfr.isna().sum().sort_values(ascending=False)

Song.duration         32
Record Label           3
AlbumName              0
Title                  0
Performer              0
Track.number           0
Composer               0
Year                   0
Genre                  0
Styles                 0
acousticness           0
danceability           0
energy                 0
instrumentalness       0
liveness               0
loudness               0
speechiness            0
tempo                  0
valence                0
duration_ms            0
popularity             0
shake.the.audience     0
romantic               0
obscene                0
family.spiritual       0
sadness                0
dtype: int64

### Ovakav način regulisanja nedostajućih vrednosti pogodniji je za nenumeričke kolone, jer kod njih nije moguće dobiti srednju vrednost kolone. Kolona `Record Label` je kategorička, stoga je kod nje najbolje primeniti metodu `dropna`.

In [6]:
dfr.dropna(subset=['Record Label', 'Song.duration'], inplace=True)

### Budući da smo otklonili sve nedostajuće vrednosti, izmenjeni dataset ćemo sačuvati u novi csv fajl.

In [7]:
dfr.to_csv('../data/brit_bez_NaN.csv', index=False)

### Sada ćemo učitati fajl koji smo sačuvali kako bismo nastavili njegovo sređivanje. Izvršićemo sređivanje tekstualnih kolona. U datafrejmu ove kolone su tipa `object`.

In [8]:
dfr2 = pd.read_csv('../data/brit_bez_NaN.csv')
dfr2.select_dtypes(include='object').sample(8)

Title               AlbumName  \
981                  Boris the Spider              A Quick One   
791                        Good Times         Out of Our Heads   
735  There Will Never Be Another Day                  Emotions   
304       Hey Gyp (Dig the Slowness)                 Fairytale   
53                It Don't Feel Good              Try Too Hard   
307                Three King Fishers   Sunshine Superman [US]   
761                      Empty Heart                    12 X 5   
17            Any Time You Want Love             American Tour   

                                          Record Label            Performer  \
981                                                MCA              The Who   
791                                      ABKCO Records   The Rolling Stones   
735                                  JVC Compact Discs    The Pretty Things   
304  Castle Music Ltd. / Fontana / Sanctuary / Sanc...              Donovan   
53                                                 BMG  The Dave Clark Five   
307                                      Legacy / Epic              Donovan   
761                                      ABKCO Records   The Rolling Stones   
17                                                 BMG  The Dave Clark Five   

    Song.duration                                     Composer  \
981          2:26                               John Entwistle   
791          1:59                                    Sam Cooke   
735          2:22  Phil May / Dick Taylor / Wally Allen Waller   
304          3:09                     Donovan / Donovan Leitch   
53           1:41                    Dave Clark / Denis Payton   
307          3:16                                      Donovan   
761          2:38                 Mick Jagger / Keith Richards   
17           2:12                  Dave Clark / Lenny Davidson   

              Genre                                             Styles  
981        Pop/Rock  Album Rock, British Invasion, Contemporary Pop...  
791        Pop/Rock  Blues-Rock, British Blues, British Invasion, R...  
735        Pop/Rock  British Invasion, Prog-Rock, Psychedelic/Garag...  
304            Folk  British Folk, British Invasion, Folk Revival, ...  
53         Pop/Rock                      British Invasion, Rock & Roll  
307  Pop/Rock, Folk  British Folk, British Invasion, British Psyche...  
761        Pop/Rock  Blues-Rock, British Blues, British Invasion, C...  
17         Pop/Rock      British Invasion, Early Pop/Rock, Rock & Roll

### Iz kolone `Composer` ćemo izdvojiti prvog kompozitora i sačuvati ga u novu kolonu `First composer`

In [9]:
def izdvoj_prvog_komp(comps):
    kl =  comps.split('/')
    return kl[0].strip()
print(izdvoj_prvog_komp('Dave Clark / Mike Smith'))

Dave Clark


### Sada ćemo ovu funkciju primeniti (funkcija `apply`) na svako polje iz kolone `Composer`, i na taj način ćemo dobiti željenu novu kolonu.

In [10]:
dfr2['First composer'] = dfr2['Composer'].apply(izdvoj_prvog_komp)
dfr2['First composer'].sample(10)

344            Donovan
411      Steve Cropper
920    Tom Springfield
820        Mick Jagger
858        Frank Allen
143        John Lennon
276       Georges Aber
989     Pete Townshend
106      James Bracken
532           Pat Best
Name: First composer, dtype: object

### Kod kolone `Performer` postoji problem da se umetnuo znak za novi red (\n) kod određenih izvođača. Rešićemo to preko funkcije strip (split u javi). 

In [11]:
dfr2['Performer'].str.contains('\n').sum()

np.int64(47)

### Primećujemo da ima 47 zapisa koji sadrže znak ove za novi red. Ovaj put nećemo pisati funkciju za otklon tih znakova pre samog poziva apply funkcije, već ćemo to uraditi direktno u njoj preko lambda izraza.

In [12]:
dfr2['Performer'] = dfr2['Performer'].apply(lambda izv: izv.replace('\n', ''))

dfr2['Performer'].str.contains('\n').sum()

np.int64(0)

### Nakon što smo rešili ovaj problem, utvrdili smo da je sem znakova za novi red, umetnuto pedesetak znakova za razmak u samom nazivu izvođača.

In [13]:
def brisi_razmake(izv):
    reci = izv.split(' ')
    words = [r for r in reci if len(r)>0]
    return ' '.join(words)

dfr2['Performer'] = dfr2['Performer'].apply(brisi_razmake)

### Nakon što smo rešili i ovaj problem, kolona Performers je ispravna, stoga možemo sačuvati promene u novi csv fajl.

In [14]:
dfr2.to_csv('../data/brit_sredjen.csv', index=False)